Add support to disable DGPU in EDK2

filipleple opened the pull request.

XutaxKamay committed:
Added support to disable DGPU in EDK2

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/clevo/mtl-h/variants/dgpu/gpio.c: add dgpu disable condition

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/clevo/mtl-h/variants/dgpu/ramstage.c: add dgpu PCI device disable logic

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
configs/novacustom_v5*0tnx: bump EDK2, enable dgpu option

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
configs/config.novacustom_*: bump EDK2 revision

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
vendorcode/dasharo/options.c: handle DGPUEnabled as uint8

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/clevo/mtl-h/variants/dgpu/ramstage.c: fix implicit uint8 to bool conversion

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
src/mainboard/clevo/mtl-h/variants/: add extra dgpu-related debug prints

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

mkopec commented on src/mainboard/clevo/mtl-h/variants/dgpu/gpio.c:
```diff
@@ -314,16 +315,21 @@ void mainboard_configure_gpios(void)
 	} else {
-		printk(BIOS_ERR, "dGPU failed to power on, turning off\n");
-		gpio_set(DGPU_PWR_EN, 0);
+			gpio_set(DGPU_PWR_EN, 0);
```
~~this is the default state (configured in gpio_configure_pads) so this else block is not necessary~~

mkopec replied:
@filipleple You might also need to set DGPU_RST_N to 0 here

mkopec commented on src/mainboard/clevo/mtl-h/variants/dgpu/gpio.c:
```diff
@@ -314,16 +315,21 @@ void mainboard_configure_gpios(void)
-		printk(BIOS_INFO, "dGPU powered on\n");
-		gpio_set(DGPU_RST_N, 1);
+	if (dasharo_is_dgpu_enabled()) {
+		/* dGPU power on sequence */
```
Try setting DGPU_RST_N to 0 here explicitly

filipleple replied:
@mkopec still no luck :( How about forcing a PCI retraining/rescanning after enabling the device in ramstage?

mkopec replied:
I don't think it should be necessary. I mean we even have runtime poweroff / poweron to save power, without retraining. Take a look at the power sequence in `src/mainboard/clevo/mtl-h/acpi/dgpu/power.asl`

```asl
/* GCOFF Out, i.e. full power-on sequence */
Method (GCOO, 0, Serialized)
{
	If (GPPS == GPU_POWER_STATE_ON)
	{
		Printf ("PGON: GPU already on")
		Return
	}

	SRCC (SRCCLK_ENABLE)
	PGON ()
	\_SB.PCI0.RP12.LD23 ()

	/* Wait for dGPU to reappear on the bus */
	Local0 = 50
	While (NVID != PCI_VID_NVIDIA)
	{
		Stall (100)
		Local0--
		If (Local0 == 0)
		{
			Break
		}
	}

	/* Restore the PEG LTR enable bit */
	LREN = SLTR

	/* Clear recoverable errors detected bit */
	CEDR = 1

	/* Restore the PEG LTR enable bit */
	LREN = SLTR

	/* Clear recoverable errors detected bit */
	CEDR = 1
}

/* GCOFF exit sequence */
Method (PGON, 0, Serialized)
{
	Local0 = Timer - GCOT
	If (Local0 < MIN_OFF_TIME_TIMERS)
	{
		Local1 = (MIN_OFF_TIME_TIMERS - Local0) / 10000
		Printf("Sleeping %o to ensure min GCOFF time", Local1)
		Sleep (Local1)
	}

	/* Assert PERST# */
	CTXS (GPIO_GPU_PERST_L)

	/* Enable power rails */
	STXS (GPIO_GPU_PWR_EN)

	/* Wait 200ms for PWRGD */
	GPPL (GPIO_GPU_PWR_GD, 1, 200)

	/* Deassert PERST# */
	STXS (GPIO_GPU_PERST_L)

	GC6E = GC6_STATE_EXITED
	GPPS = GPU_POWER_STATE_ON
}
```

I think the logic around SRCCLK might have something to do with this.

mkopec replied:
oh and maybe also move mainboard_configure_gpios() from just before FSP-M in romstage to mainboard_init in ramstage. FSP-M might be messing with our config

mkopec commented on src/mainboard/clevo/mtl-h/variants/dgpu/gpio.c:
```diff
@@ -310,20 +318,69 @@ static const struct pad_config gpio_table[] = {
+        mdelay(10);  // Required delay before power-on
+
+        /* Enable PCIe SRCCLK */
+        p2sb_acpi_set_pci_clock(GPU_SRCCLK_PIN, true);
```
this just generates ACPI bytecode, it doesn't actually directly control the srcclk

filipleple commented on src/soc/intel/meteorlake/fsp_params.c:
```diff
@@ -647,6 +647,7 @@ static void fill_fsps_pcie_params(FSP_S_CONFIG *s_cfg,
 		s_cfg->PcieRpHotPlug[i] = !!(rp_cfg->flags & PCIE_RP_HOTPLUG)
 				|| CONFIG(SOC_INTEL_COMPLIANCE_TEST_MODE);
 		s_cfg->PcieRpClkReqDetect[i] = !!(rp_cfg->flags & PCIE_RP_CLK_REQ_DETECT);
+		s_cfg->PcieRpDetectTimeoutMs[i] = rp_cfg->pcie_rp_detect_timeout_ms;
```
@mkopec setting the PcieRpDetectTimeoutMs like in adl fsp_params, didn't help

mkopec replied:
did you set the timeout to some value in the devicetree? Here you've only hooked up the FSP param to the devicetree. You need something like `register "pcie_rp_detect_timeout_ms" = "10000"` right about here: https://github.com/Dasharo/coreboot/blob/ffb2e08621a86339dc86a94155a8740fef5377e3/src/mainboard/clevo/mtl-h/variants/dgpu/overridetree.cb#L17-L21

@filipleple

filipleple replied:
[yep](https://github.com/Dasharo/coreboot/pull/630/files#diff-fbcfd0fafdab6c8b4438f49902e708697ffc2452a18b982ff3992710a81e84c8R21), still no luck

mkopec replied:
ugh I'm running out of ideas at this point. Did you try with debug fsp already?

filipleple commented on src/mainboard/clevo/mtl-h/variants/dgpu/gpio.c:
```diff
@@ -331,10 +346,10 @@ void mainboard_configure_gpios(void)
-        printk(BIOS_DEBUG, "Enabled PCIe SRCCLK for dGPU\n");
-
+        // p2sb_dev_unhide(dev);
+        check_gpu_clock_state();
```
Always returns `GPU Clock PCR_BIOS_BUFFEN (PID_ISCLK) = 0x0000003F`, whether dGPU does or doesn't get detected

mkopec committed:
mb/clevo/mtl-h: Fix dGPU not coming up after reboot

1. Don't try to power it on again if it was powered on already.
2. Kick the platform via global reset if the GPU fails to come up. This
   reliably fixes dGPU detection.

Change-Id: I76b80ec8c988d25adff23f02b697d98885848f35
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented on src/mainboard/clevo/mtl-h/variants/dgpu/overridetree.cb:
```diff
@@ -17,7 +17,7 @@ chip soc/intel/meteorlake
 				.clk_src = 6,
 				.clk_req = 6,
-				.flags = PCIE_RP_LTR | PCIE_RP_AER | PCIE_RP_CLK_REQ_DETECT,
+				.flags = PCIE_RP_LTR | PCIE_RP_HOTPLUG,
```
This change is probably no longer needed, forgot to remove

filipleple replied:
@mkopec nope :sweat_smile:  we need this one as well, oddly enough. Tested a couple times to be sure

filipleple committed:
configs/novacustom: bump EDK2 revision on dGPU laptops

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

mkopec commented on src/mainboard/clevo/mtl-h/variants/dgpu/ramstage.c:
```diff
@@ -2,10 +2,41 @@
 	i2c_amp_dev->enabled = have_smartamp;
+
+	struct device *dgpu_dev = pcidev_on_root(0x01, 0);
+	dgpu_dev->enabled = dasharo_is_dgpu_enabled() != 0;
+	printk(BIOS_DEBUG, "dgpu_dev->enabled: %d\n", dgpu_dev->enabled);
```
possible null pointer deref

```suggestion
	struct device *dgpu_dev = pcidev_on_root(0x01, 0);
	if (dgpu_dev) {
	        dgpu_dev->enabled = dasharo_is_dgpu_enabled() != 0;
	        printk(BIOS_DEBUG, "dgpu_dev->enabled: %d\n", dgpu_dev->enabled);    
	}
```

mkopec replied:
Same applies to the i2c amp dev above actually, can you fix that one as well?

filipleple replied:
[Done](https://github.com/Dasharo/coreboot/pull/630/commits/fe4cc24a49b692d4358aa8d196513a478d93f995)

mkopec reviewed (changes requested):
misclick

filipleple committed:
mainboard/clevo/mtl-h/variants/dgpu/ramstage.c: fix possible null pointer dereference

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

mkopec commented on configs/config.novacustom_nv4x_tgl_dgpu:
I'm not too happy about adding another variant of the board. CFR could help hide the option on igpu models without having to build a second binary for it
